teps/0077-pipeline-partial-execution.md

```diff
@@ -0,0 +1,496 @@
+
+Together these will bring in the ability to resume/retry a failed `PipelineRun`
+
+## Motivation
+
+1. A very long Pipeline may fail due to a transient failure, and the user may
+   want to only rerun the `Tasks` that failed. This is the most significant
+   motivation, as we use tekton as a backend for running our machine learning
+   pipelines at Kubeflow.
+2. It is not enough to `retry` a `PipelineTask` n times, as the failures can
+   be due to e.g. service outage. A manual resume/ retry may be helpful.
+3. Iterate quickly, by disabling tasks that take longer time. This can be done
+   at run time, i.e. without editing pipeline definition.
+
+### Goals
+
+- Partially execute a `Pipeline`, by disabling tasks.
+- Resume a failed or cancelled `PipelineRun` from previous `PipelineRun`.
+- Discuss how the results/workspaces of disabled tasks can be referenced or
+  populated from previous `PipelineRun`.
+
+### Non-Goals
+
+- TBD.
+
+### Use Cases (optional)
+
+1. *Optimal use of resources*: `tektoncd` as a backend for ML. 
+   A machine learning pipeline may consist of tasks moving large amount of
+   data and then training ml models, all of it can be very resource consuming
+   and inability to retry would require a user to start the entire pipeline
+   over. A manual retry, with the ability to specify what tasks should
+   be skipped, may be helpful.
+2. Partial execution of pipeline. This is useful for reusing an existing
+   pipeline i.e. a user can disable certain task from an existing pipeline
+   and in this way run it without creating a new pipeline.
+3. Partial execution is also helpful for testing, i.e. skipping some tasks
+   and developing and testing iteratively and quickly.
+4. Pause and resume, i.e. one could manually cancel a running `PipelineRun`
+   and resume at later point.
+
+## Requirements
+
+- Create a new `PipelineRun` to resume or retry a completed `PipelineRun`.
+- Ability to partially run a Pipeline, without editing its definition or without
+  an existing `PipelineRun`.
+- An end user may not have to figure out the execution DAG, to be able to resume
+  or retry.
+
+## Proposal
+
+### Requesting API Changes
+1. Add `pipelineRunRef` under `PipelineRun.spec`. It has following fields: 
+   - `pipelineRunRef.name` which is the name of previously run `PipelineRun`.
+   - `pipelineRunRef.enableTasks` accepts an array of task names under it.
+2. Add `disableTasks` under `PipelineRun.spec`, which accepts an array
+    of task name.
+   - `name`: Name of the task to be disabled.
+
+Q. Why do we need `pipelineRunRef` when we have `disableTasks`? 
+
+`disableTasks` can be used to explicitly disable tasks that a user 
+do not wish to run. On the other hand, in `pipelineRunRef` tekton controller
+automatically figures out the tasks failed and unfinished, because it knows the
+DAG. For the end user, it can be difficult to figure out the DAG and prepare
+the accurate execution plan for the next pipeline run.
```

Comment:
I'd think that for the end user, they need to figure out the DAG so that they can decide what to `enable` and `disable` -- I'm worried about the automaton of executing `failed` and `unfinished` tasks, what if a user wants to also include `skipped` tasks? does `failed` include `cancelled` tasks?

maybe the smallest change we can make is introducing the `disabled` tasks (with a different name) to solve most of the use cases, then explore the automation as the next step -- what do you think?

Reply:
i definitely agree with @jerop here about introducing just the`disabletasks` functionality here and tackling `pipelineRunRef` in a subsqeuent proposal

This would mean making "Resume a failed or cancelled `PipelineRun` from previous `PipelineRun`" a non-goal of this TEP - to me including this goal and the `pipelineRunRef` functionality in this TEP makes the line between [TEP-0065](https://github.com/tektoncd/community/pull/422)  very blurry

i also think it'll be easier to get this TEP approved more quickly if we focus on just the `disabledTasks` functionality first

Reply:
Ok, what if a user wants to resume a PipelineRun which has pipelineSpec embedded in it?

Reply:
I think the comments here show that the various use cases, running a partial pipeline, resuming a cancelled pipeline and retrying a failed pipeline come with different priorities and requirements. 

The initial TEP started by @ScrapCodes was about retrying a failed taskrun/pipelinerun. 
We saw common aspects with the partial execution which lead to this TEP, but perhaps we should first do some work to spell out what are the common aspects between the various use cases / features.

In my mind I see re-running a failed / cancelled pipeline - without redoing all the work - as the main use case. For that to be useful it should be up to the controller to decide what to run, not to the user.

The partial execution use case has different needs. The user wants to skip an expensive task in the pipeline for testing purposes or because a system it depends on is temporarily not available. In this case it's up to the user to say which tasks to skip, and it's up to the controller to say if the request can be served and what are the consequences - i.e. other tasks might be skipped because of the missing ones.